### PR TITLE
feat(sca): change sca packages output to include dependencies structure

### PR DESCRIPTION
**Pull request by NoaAzoulay**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

for files without dependencies the output remains the same .
for files with dependencies the output change to :

Fixes # (issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/common/sca/output.py [hunk 1]
```diff
@@ -293,33 +297,122 @@ def add_to_reports_cves_and_packages(
...
+        sca_details: SCADetails | None = None,
+        report_type: str | None = None,
 ) -> None:
     licenses_per_package_map: dict[str, list[str]] = \
         _add_to_report_licenses_statuses(report, check_class, scanned_file_path, rootless_file_path, runner_filter,
                                          license_statuses, sca_details, report_type)
 
-    add_to_reports_cves_and_packages(report, check_class, scanned_file_path, rootless_file_path, runner_filter,
-                                     vulnerabilities, packages, licenses_per_package_map, sca_details, report_type,
-                                     ScanDataFormat.TWISTCLI if dependencies is None else ScanDataFormat.DEPENDENCY_TREE)
```
Comment (ayajbara) [inline checkov/common/sca/output.py:322]: why did you change the `scan_data_format`?

File: checkov/common/sca/output.py [hunk 2]
```diff
@@ -238,52 +250,44 @@ def _add_to_report_licenses_statuses(
 
 
 def add_to_reports_cves_and_packages(
```
Comment (ayajbara) [inline checkov/common/sca/output.py:252]: huge function, can we split it?

File: checkov/common/sca/output.py [hunk 3]
```diff
@@ -83,12 +86,12 @@ def create_report_license_record(
...
-    details: dict[str, Any],
-    vulnerability_details: dict[str, Any],
-    sca_details: SCADetails | None = None,
-    scan_data_format: ScanDataFormat = ScanDataFormat.TWISTCLI
+        details: dict[str, Any],
+        vulnerability_details: dict[str, Any],
+        sca_details: SCADetails | None = None,
+        scan_data_format: ScanDataFormat = ScanDataFormat.TWISTCLI
 ) -> None:
-    if scan_data_format in {ScanDataFormat.TWISTCLI, ScanDataFormat.DEPENDENCY_TREE}:
```
Comment (ayajbara) [inline checkov/common/sca/output.py:91]: also here, why you change it?

File: checkov/common/sca/output.py [hunk 4]
```diff
@@ -160,8 +166,12 @@ def create_report_cve_record(
         "description": description,
         "risk_factors": vulnerability_details.get("riskFactors"),
         "published_date": vulnerability_details.get("publishedDate")
-        or (datetime.now() - timedelta(days=vulnerability_details.get("publishedDays", 0))).isoformat(),
+                          or (datetime.now() - timedelta(
+            days=vulnerability_details.get("publishedDays", 0))).isoformat(),
         "licenses": licenses,
+        "root_package_alias": root_package_alias,
```
Comment (ayajbara) [inline checkov/common/sca/output.py:172]: adding `root_package_alias` to every record will add an extra space that maybe in the future we will need it for other things, when we can create it when we need it using the package name and version.

File: checkov/common/sca/output.py [hunk 5]
```diff
@@ -238,52 +250,44 @@ def _add_to_report_licenses_statuses(
...
-        report.add_record(cve_record)
-        vulnerable_packages.append(get_package_alias(package_name, package_version))
+        vulnerable_packages[package_alias].append(vulnerability)
 
     for package in packages:
-        if get_package_alias(package["name"], package["version"]) not in vulnerable_packages:
+        package_alias = get_package_alias(package["name"], package["version"])
+
+        if dependencies:
+            if "root" in package and package["root"]:
```
Comment (ayajbara) [inline checkov/common/sca/output.py:281]:
```suggestion
            if package.get("root", ""):
```

File: checkov/common/sca/output.py [hunk 6]
```diff
@@ -293,33 +297,122 @@ def add_to_reports_cves_and_packages(
...
                         "package_type": get_package_type(package["name"], package["version"], sca_details),
                     },
                 )
             )
 
+    for root_package_index in root_packages_list:
+        deps = []
+        if dependencies:
+            visited: set[int] = set()  # Set to keep track of visited nodes.
+            dfs(visited=visited, graph=dependencies, root=root_package_index)
```
Comment (ayajbara) [inline checkov/common/sca/output.py:311]: by visited do you mean root_package_dependencies?

File: checkov/common/sca/output.py [hunk 7]
```diff
@@ -293,33 +297,122 @@ def add_to_reports_cves_and_packages(
...
-                        "licenses": format_licenses_to_string(licenses_per_package_map[get_package_alias(package["name"], package["version"])]),
+                        "licenses": format_licenses_to_string(
+                            licenses_per_package_map[package_alias]),
                         "package_type": get_package_type(package["name"], package["version"], sca_details),
                     },
                 )
             )
 
+    for root_package_index in root_packages_list:
+        deps = []
```
Comment (ayajbara) [inline checkov/common/sca/output.py:308]:
```suggestion
        vulnerable_dependencies = []
```
since its not all the dependencies, only vulnerable dependencies

File: checkov/common/sca/output.py [hunk 8]
```diff
@@ -293,33 +297,122 @@ def add_to_reports_cves_and_packages(
...
+            add_cve_record_to_report(vulnerability_details=cve, package=root_package,
+                                     root_package_alias=root_package_alias,
+                                     rootless_file_path=rootless_file_path, scanned_file_path=scanned_file_path,
+                                     check_class=check_class, licenses_per_package_map=licenses_per_package_map,
+                                     runner_filter=runner_filter, sca_details=sca_details,
+                                     scan_data_format=scan_data_format, report_type=report_type, report=report,
+                                     root_package_version=root_package_version,
+                                     root_package_name=root_package_name)
+        if "deps" in root_package:
+            for dep in root_package["deps"]:
```
Comment (ayajbara) [inline checkov/common/sca/output.py:335]:
```suggestion
       for dep in root_package.get("deps", []):
```

File: checkov/common/sca/output.py [hunk 9]
```diff
@@ -111,14 +114,17 @@ def _update_details_by_scan_data_format(
...
-    sca_details: SCADetails | None = None,
-    scan_data_format: ScanDataFormat = ScanDataFormat.TWISTCLI
+        rootless_file_path: str,
+        file_abs_path: str,
+        check_class: str,
+        vulnerability_details: dict[str, Any],
+        licenses: str,
+        root_package_alias: str,
+        root_package_version: str,
+        root_package_name: str,
```
Comment (ayajbara) [inline checkov/common/sca/output.py:124]:
```suggestion
        root_package_alias: str| None = None,
        root_package_version: str| None = None,
        root_package_name: str| None = None,
```
twistlock scan for example didnt have dependencies and these fields should be an optional for this kind of scan

File: checkov/common/sca/output.py (hunk 2 above)
Comment (ayajbara) [inline checkov/common/sca/output.py:252]: why not to create a new function with all the changes for dependency tree scan results and to leave the old one for twistlock? 
the old one faster and when we dont have any dependencies we dont really need to make all these steps to add a record.

**Review by ayajbara — COMMENTED** [review]
GJ 💯 added a few comments and please add tests for scan results with and without dependences, for the two scan types dependency tree and twistlock :)

File: checkov/common/sca/output.py (hunk 6 above)
Comment (NoaAzoulay) [reply checkov/common/sca/output.py:311]: yes, it's for the dfs, i'll change the name

File: checkov/common/sca/output.py (hunk 2 above)
Comment (NoaAzoulay) [reply checkov/common/sca/output.py:252]: we wanted to keep the same flow for all the cases, @Saarett @ayajbara @matansha WDT?

File: checkov/sca_package_2/output.py [hunk 10]
```diff
@@ -86,20 +86,22 @@ def create_cli_output(fixable: bool = True, *cve_records: list[Record]) -> str:
             #  this shouldn't happen
             logging.error(f"'vulnerability_details' is not set for {record.check_id}")
             continue
+        package_alias = get_package_alias(
+            record.vulnerability_details.get("root_package_name", record.vulnerability_details["package_name"]),
+            record.vulnerability_details.get("root_package_version", record.vulnerability_details["package_version"])
+        )
```
Comment (itai1357) [inline checkov/sca_package_2/output.py:92]: 1. maybe rename 'package_alias' to 'root_package_alias'
this way it will be a little bit clearer that we are grouping by the root packages.
2. is it possible that root_package_name is empty? what happen in case of the package is root ? I think that both package_name and root_package_name should be not empty and equal (the same for package_version). If I'm right, is is enough to have
```
record.vulnerability_details.get["root_package_name"],
```
unless I'm missing anything

File: checkov/sca_package_2/output.py [hunk 11]
```diff
@@ -120,38 +122,41 @@ def create_cli_output(fixable: bool = True, *cve_records: list[Record]) -> str:
                     severity_str = record.severity.name.lower() if record.severity else BcSeverities.NONE.lower()
                     setattr(cve_count, severity_str, getattr(cve_count, severity_str) + 1)
 
+                    package_name = record.vulnerability_details["package_name"]
+                    package_version = record.vulnerability_details["package_version"]
+
                     if record.vulnerability_details["lowest_fixed_version"] != UNFIXABLE_VERSION:
                         cve_count.has_fix += 1
+                    if root_package_alias == get_package_alias(package_name, package_version):
```
Comment (itai1357) [inline checkov/sca_package_2/output.py:130]:
```suggestion
                    is_root_package = root_package_alias == get_package_alias(package_name, package_version)
                    if is_root_package:
```

File: checkov/sca_package_2/output.py [hunk 12]
```diff
@@ -120,38 +122,41 @@ def create_cli_output(fixable: bool = True, *cve_records: list[Record]) -> str:
...
                         }
                     )
                 elif record.check_name == SCA_LICENSE_CHECK_NAME:
                     if record.check_result["result"] == CheckResult.SKIPPED:
                         continue
                     should_print_licenses_table = True
-                    package_licenses_details_map[package_name].append(
-                        _LicenseStatus(package_name=package_name,
+                    package_licenses_details_map[record.vulnerability_details["package_name"]].append(
+                        _LicenseStatus(package_name=record.vulnerability_details["package_name"],
```
Comment (itai1357) [inline checkov/sca_package_2/output.py:149]: notice that record.vulnerability_details["package_name"]==package_name so:
```suggestion
                    package_licenses_details_map[package_name].append(
                        _LicenseStatus(package_name=package_name,
```

File: checkov/sca_package_2/output.py [hunk 13]
```diff
@@ -120,38 +122,41 @@ def create_cli_output(fixable: bool = True, *cve_records: list[Record]) -> str:
...
                                        package_version=record.vulnerability_details["package_version"],
```
Comment (itai1357) [inline checkov/sca_package_2/output.py:150]:
```suggestion
                                       package_version=package_version,
```

File: checkov/sca_package_2/output.py (hunk 10 above)
Comment (NoaAzoulay) [reply checkov/sca_package_2/output.py:92]: @itai1357 
1. sure, i'll rename it
2. in case the cve is on a package which is a root so we'll have the same value in "package_name" and "root_package_name", and the same value for "package_version" and "root_package_version". we are using this fact in the creation of the cve table lines. so "root_package_name" or "root_package_version" won't be empty anyway

**Review by itai1357 — APPROVED** [review]

**Review by ayajbara — APPROVED** [review]

File: checkov/sca_image/runner.py [hunk 14]
```diff
@@ -24,7 +24,7 @@
 from checkov.common.util.dockerfile import is_docker_file
 from checkov.common.util.http_utils import request_wrapper
 from checkov.runner_filter import RunnerFilter
-from checkov.sca_package.runner import Runner as PackageRunner
+from checkov.sca_package_2.runner import Runner as PackageRunner
```
Comment (ayajbara) [inline checkov/sca_image/runner.py:27]: it will break a lot of things.
